extensions/binding/org.eclipse.smarthome.binding.openweathermap/src/main/java/org/eclipse/smarthome/binding/openweathermap/internal/connection/OpenWeatherMapConnection.java

```diff
@@ -0,0 +1,244 @@
+import org.eclipse.smarthome.binding.openweathermap.internal.model.OpenWeatherMapJsonHourlyForecastData;
+import org.eclipse.smarthome.binding.openweathermap.internal.model.OpenWeatherMapJsonWeatherData;
+import org.eclipse.smarthome.core.cache.ExpiringCacheMap;
+import org.eclipse.smarthome.core.library.types.RawType;
+import org.eclipse.smarthome.io.net.http.HttpUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import com.google.gson.JsonSyntaxException;
+
+/**
+ * The {@link OpenWeatherMapConnection} is responsible for handling the connections to OpenWeatherMap API.
+ *
+ * @author Christoph Weitkamp - Initial contribution
+ */
+@NonNullByDefault
+public class OpenWeatherMapConnection {
+
+    private final Logger logger = LoggerFactory.getLogger(OpenWeatherMapConnection.class);
+
+    private static final String UTF_8_ENCODING = "UTF-8";
+
+    private static final String PARAM_APPID = "appid";
+    private static final String PARAM_UNITS = "units";
+    private static final String PARAM_LAT = "lat";
+    private static final String PARAM_LON = "lon";
+    private static final String PARAM_LANG = "lang";
+    private static final String PARAM_FORECAST_CNT = "cnt";
+
+    // Current weather data (see https://openweathermap.org/current)
+    private static final String WEATHER_URL = "https://api.openweathermap.org/data/2.5/weather";
+    // 5 day / 3 hour forecast (see https://openweathermap.org/forecast5)
+    private static final String THREE_HOUR_FORECAST_URL = "https://api.openweathermap.org/data/2.5/forecast";
+    // 16 day / daily forecast (see https://openweathermap.org/forecast16)
+    private static final String DAILY_FORECAST_URL = "https://api.openweathermap.org/data/2.5/forecast/daily";
+    // Weather icons (see https://openweathermap.org/weather-conditions)
+    private static final String ICON_URL = "https://openweathermap.org/img/w/";
+
+    private final OpenWeatherMapAPIHandler handler;
+    private final HttpClient httpClient;
+
+    private static final ExpiringCacheMap<String, @Nullable RawType> IMAGE_CACHE = new ExpiringCacheMap<>(
+            TimeUnit.DAYS.toMillis(7));
+    private final ExpiringCacheMap<String, String> cache;
+
+    private final JsonParser parser = new JsonParser();
+    private final Gson gson = new Gson();
+
+    public OpenWeatherMapConnection(OpenWeatherMapAPIHandler handler, HttpClient httpClient) {
+        this.handler = handler;
+        this.httpClient = httpClient;
+
+        OpenWeatherMapAPIConfiguration config = handler.getOpenWeatherMapAPIConfig();
+        cache = new ExpiringCacheMap<>(TimeUnit.MINUTES.toMillis(config.getRefreshIntervall()));
+    }
+
+    @Nullable
+    public synchronized OpenWeatherMapJsonWeatherData getWeatherData(OpenWeatherMapLocationConfiguration locationConfig)
```

Comment:
Please annotate the result type with `@Nullable` instead (see [Null Annotations Conventions](https://www.eclipse.org/smarthome/documentation/development/conventions.html#null-annotations)).